### PR TITLE
Add truffle_import_cached in truffle.h

### PR DESCRIPTION
**Pull request by grimmerm**
This function allows us to import a foreign value (see also truffle_import function).
In contrast to truffle_import,  truffle_import_cached caches the imported value internally.
This caching guarantees good performance because we save the foreign value look-up.

Truffle_import_cached allows you to import a foreign value and keep is as a function local (rather than a global variable) without a loss in performance (which would be introduced by importing a foreign value every time truffle_import is called). 

For example:

Rather than writing 
```C
static void *ruby_cext;

__attribute__((constructor))
void truffle_ruby_load() {
  ruby_cext = truffle_import("ruby_cext");
}
```

one can now do

```C
__attribute__((constructor))
void truffle_ruby_load() {
  void *ruby_cext = truffle_import_cached("ruby_cext");
}
```
.

This should solve #196.

### CODE REVIEW

**Comment by chrisseaton** [comment]
I don't see the need for a separate name for the cached variant. Why not make all imports cached?

File: projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/interop/LLVMTruffleImportCached.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            if (this.name.equals(n)) {
+                return value;
+            } else {
+                return next.execute(n);
+            }
+        }
+
+    }
+
+}
```
Comment (chrisseaton) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/interop/LLVMTruffleImportCached.java]: Could this not be implemented using the DSL's `@Cached`?

File: projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/interop/LLVMTruffleImportCached.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    private static final class ResolvedCacheEntry extends CacheEntry {
+
+        private final String name;
+        private final LLVMTruffleObject value;
+        @Child private CacheEntry next;
+
+        public ResolvedCacheEntry(String name, LLVMTruffleObject value, CacheEntry next) {
+            this.name = name;
+            this.value = value;
+            this.next = insert(next);
```
Comment (woess) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/interop/LLVMTruffleImportCached.java]: you shouldn't have to insert in the constructor

File: projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/interop/LLVMTruffleImportCached.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            TruffleObject foreignObject = (TruffleObject) LLVMLanguage.INSTANCE.getEnvironment().importSymbol(name);
+            LLVMTruffleObject value = new LLVMTruffleObject(foreignObject);
+
+            Node n = this;
+            while (!(n instanceof LLVMTruffleImportCached)) {
+                n = n.getParent();
+            }
+            LLVMTruffleImportCached head = (LLVMTruffleImportCached) n;
+            head.cache = insert(new ResolvedCacheEntry(name, value, head.cache));
+            head.adoptChildren();
```
Comment (woess) [inline projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/interop/LLVMTruffleImportCached.java]: I believe this should be `head.insert(...)` - probably the reason why you added `adoptChildren`?

**Comment by grimmerm** [comment]
@chrisseaton, the thing with always caching is that we would miss an updated exported symbol.

e.g.
JS: export("foo", a)
C: import("foo") // imports a
JS: export("foo", b)
C: import("foo") // would return the cached a object

do we really want that?

**Comment by chrisseaton** [comment]
We should probably implement invalidation within the Truffle API. When you import an object, you could get the object but also an `Assumption` which is invalidated when that name is replaced with a new object. The Truffle API would manage that invalidation.

**Pull request merged**